src/schema/devcards.ts

```diff
@@ -0,0 +1,185 @@
+
+    """
+    Timestamp the dev card was created
+    """
+    createdAt: DateTime!
+
+    """
+    Theme of the dev card
+    """
+    theme: String!
+
+    """
+    Whether the dev card cover is from the profile, default false
+    """
+    isProfileCover: Boolean!
+
+    """
+    Whether the dev card has a border, default true
+    """
+    showBorder: Boolean!
+
+    """
+    User reputation
+    """
+    reputation: Int!
+
+    """
+    Number of articles read
+    """
+    articlesRead: Int!
+
+    """
+    Tags
+    """
+    tags: [String!]!
+
+    """
+    Sources logo image URLs
+    """
+    sourcesLogos: [String!]!
+  }
+
+  """
+  The type returned on generateDevCard mutation
+  """
+  type DevCardImage {
+    imageUrl: String!
+  }
+
+  extend type Query {
+    """
+    Get devcard by id
+    """
+    devCard(
+      """
+      Id of the requested devcard
+      """
+      id: ID
+    ): DevCardData!
+  }
+
+  extend type Mutation {
+    """
+    Generates or updates the user's Dev Card preferences
+    """
+    generateDevCard(
+      theme: DevCardTheme
+      type: DevCardType
+      isProfileCover: Boolean
+      showBorder: Boolean
+    ): DevCardImage @auth
+  }
+`;
+
+interface GenerateDevCardInput {
+  theme: DevCard['theme'];
+  isProfileCover: DevCard['isProfileCover'];
+  showBorder: DevCard['showBorder'];
+  type: 'DEFAULT' | 'WIDE' | 'X';
+}
+
+interface DevCardByIdResult extends DevCard, DevCardData {}
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+export const resolvers: IResolvers<any, Context> = {
+  Query: traceResolverObject({
+    devCard: async (
+      source,
+      { id }: { id: string },
+      ctx: Context,
+    ): Promise<DevCardByIdResult> => {
+      const res = await ctx.con.getRepository(DevCard).findOneBy({ id });
+      if (res !== null) {
+        const data = await getDevCardData(res.userId, ctx.con);
+        return { ...res, ...data };
+      }
+      throw new NotFoundError('DevCard not found');
+    },
+  }),
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  Mutation: traceResolverObject<any, any>({
+    generateDevCard: async (
+      source,
+      { theme, isProfileCover, showBorder, type }: GenerateDevCardInput,
+      ctx: Context,
+    ): Promise<{ imageUrl: string }> => {
+      const repo = ctx.con.getRepository(DevCard);
+      let devCard: DevCard = await repo.findOneBy({ userId: ctx.userId });
+      if (!devCard) {
+        devCard = await repo.save({
+          userId: ctx.userId,
+          theme: theme?.toLocaleLowerCase() as DevCardTheme | undefined,
+          isProfileCover,
+          showBorder,
+        });
+      }
+
+      // Avoid caching issues when devcard is generated again
+      const randomStr = Math.random().toString(36).substring(2, 5);
+      const queryStr = qs.stringify({
+        type,
+        r: randomStr,
+      });
+
+      return {
+        imageUrl: `${process.env.OG_URL}/devcards/${devCard.id}.png?${queryStr}`,
```

Comment:
Probably neater to use URL builder + searchParams, but non blocking

Reply:
I don't think you need omitBy and isEmpty. We just check if type is defined and then add to URL.searchParams with append. 

```js
const url = new URL(`/devcards/${devCard.id}.png`, process.env.OG_URL)
url.searchParams.append('r', randomStr)

if (type) {
  url.searchParams.append('type', type)
}

return { imageUrl: url.toString() };
```